### ISSUE TITLE
Mobs: Animals and Monsters

### ISSUE THREAD

**Issue by tobyplowy** [minetest/minetest_game#1253]
I think Minetest Game should have mobs or animals (because playing survival without mobs is kind of boring). There is also no sense of danger or fear without mobs, the only way you can die is fall damage, cactus or lava. Also, (This could be really good for a proper mob API because the ones mods use are laggy, or are filled with unnecessary features.)

Also I think there should be a setting for mobs if people don't want them or if their systems cannot handle them.

Thanks to @C1ffisme for The help on improving my Grammar :P

**Comment by NathanSalapat** [comment]
I've never noticed any issues with Tenplus1's mob api, and I've done some pretty crazy stuff with it. I once spawned a bunch of mobs, that were 3mb each, the actual model file size, they had animations, and I didn't have any issues with lag at all.

If you want to write us a proper mob API that fits everybody's use cases, and solves all the problems that have been outlines on other issues like this please do, but I feel it is very unlikely this will be picked up by someone when it is very controversial.

**Comment by paramat** [comment]
Mobs are optional because not everyome wants the danger and combat.
The occasional peaceful mob might be okay.

**Comment by tobyplowy** [comment]
@paramat you could have a danger setting like Minecraft Does ( it's used to set the level of hardness or completely remove monsters and only get animals or Nothing at All)

**Comment by rubenwardy** [comment]
Mobs make a world feel alive. Maybe if we do the splitting MTG in half thing, they could go in the completer one

**Comment by DonBatman** [comment]
Could have 3 different subgames. 1 - the existing game, 2 - existing with a few peaceful mobs and 3 - existing with both peaceful and dangerous mobs.
Would only have to maintain 1 base game, mobs api, peaceful mobs mod and dangerous mobs mod.

**Comment by tobyplowy** [comment]
@DonBatman I don't think the sub game thing is a good idea ( I think it should be a difficulty setting so you can change at anytime) also the thing with the sub games is a problem because if you have a world save and go to another sub game you can't play that world safe on that sub game only on the one you played previously ( so if it's too easy or too hard you can't change the difficulty without making a compleatle new world)

**Comment by tobyplowy** [comment]
@rubenwardy what do you mean by splitting in half ( making 2 sub games?) something like a minetest game+? (which has more content than the normal game) And the normal minetest game?

**Comment by rubenwardy** [comment]
Into Minetest Game and Minetest Foundation - the former tries to be a real game, whilst the latter tries to be a fairly minimalistic base for modding. Each of those games would straddle the current in terms of content - for example, mushrooms are weird in a minimal game

**Comment by tobyplowy** [comment]
@rubenwardy sorry but I still don't really understand ( the only thing I understand was that minetest game was supposed to be a real game but newer versions are focused on just a template why?) why don't make minetest game an actual game we are not restricted by anything this is not Minecraft this is minetest, (also I think if the base game is an actual game the quality of mods will improve)

**Comment by rubenwardy** [comment]
...because people still want a base for modding. This way we get two games: one a real game, another a base for modding

**Comment by MarkuBu** [comment]
With a simple on/off switch we wouldn't need two games

**Comment by tobyplowy** [comment]
@rubenwardy We meer a real game so +1 for me

**Comment by paramat** [comment]
Partly a duplicate of #515

**Comment by C1ffisme** [comment]
@paramat It's more of an issue that's in more detail on a certain subject discussed in that one. Although, this is a closer duplicate of a feature idea I made in the engine a while back.

**Comment by paramat** [comment]
Yeah, changed my wording, this issue is valid as it goes into more detail.

**Comment by tobyplowy** [comment]
@paramat what do you think about animals/monsters (also I think there should be a difficulty setting for Monsters if people don't want them)

**Comment by paramat** [comment]
Fine by me if they're optional and well designed. Hostile mobs should be off by default, it is intentional that MTGame doesn't have hostile mobs by default.
There are a lot of bad-looking mobs, and mobs that are too high-resolution and too detailed to be suitable, they need to be consistent in design with the player model. Simple, boxy and fairly cartoony and cute (monsters included).

**Comment by tobyplowy** [comment]
@paramat also texturing order is terrible (like in Minecraft textures are orderd a serten way so it's easy To edit them with texture packs) with Minetest they just use a repeating Texture and drag The textring All Around the Texture so they are inpossible to edit, and even if you do manage to edit it, hafe of the Faces use the same pixels so the Textures are messed Up :/ (for example the Arms could Use the same pixels as the legs) so please Make sure Every Polygon/bodypart has it's own Texture and the Texture is ordered
And everything that's not used by the model should not have a texture but instead be transparent or be a solid color.

**Comment by tobyplowy** [comment]
@paramat Textures shuld be ordered In This Way, because it's really easy to edit http://i.imgur.com/xDbpV6P.png

**Comment by NathanSalapat** [comment]
@tobyplowy It makes sense to use the same bit of the texture in multiple areas on the model if they are going to be the same, it's a waste of data to make four copies of the same part of the texture for each leg on a sheep when they will all have the same texture.
If the body parts need different textures I assume most modelers will create mappings with that, if they don't need different textures it just makes sense to layer them. Editing the texture is still easy, and even easier if the modder includes the UV map. ;)

**Comment by tobyplowy** [comment]
@NathanSalapat yes but if not every side of the cube has its own texture that's quite annoying (this is only a problem with texture packs) for example if you want a specific texture on the back or front or maybe the left side of the leg it's impossible because if you change one texture they all change! and that's why every cube neads its own texture for every side

**Comment by tobyplowy** [comment]
@NathanSalapat if it's so simple here download [this](https://forum.minetest.net/viewtopic.php?f=9&t=9301) mod and take the texture for the monster and try to texture that (have fun figuring out which side is what) this is an absolute nightmare to edit and almost every minetest mod uses this way of texturing mobs and its terrible

**Comment by NathanSalapat** [comment]
@tobyplowy This is why people should include, or link to a .blend file, and include a UV map. I've never had a case where having each leg textured the same was a problem.
Using a map like you linked makes lots of excess file size and isn't optimal as far as UV mapping goes. But of course you are welcome to do as you please. :)

**Comment by tobyplowy** [comment]
@NathanSalapat you know that would work but I don't want to use the model as a reference To texture (if this would have a proper organized system this would have been really easy just using the texture)

**Comment by ExeterDad** [comment]
@tobyplowy what you are requesting just can't happen when creators are coming up with different shaped models. @NathanSalapat is correct, the sources need to be made available for proper UV map editing. How the maps are created depends on the creators goals. If the goal is to have a optimized low res UV, it's common for image parts to be reused in many parts of the models.

**Comment by Ferk** [comment]
@tobyplowy if you don't want to check on the model, you are gonna need some sort of help when texturing it.
I don't think including textures for each of the 8 legs of a spider is gonna help you see them better when the low resolution and hair might make it all look like a grey brownish mess with eyes.

I understand that this could be good for those models who really deserve the attention, like the player model, which people want to be able to customize it as much as possible to make the texture their own. But I don't think there's such need in the case of monsters.

**Comment by tobyplowy** [comment]
@Ferk I think you're misunderstanding me, I'm not asking for Different image files for textures I'm asking for textures Every Side in one image for everybody part (Look at Minecraft textures for reference every side of every body part has its own individual texture all in one picture)

**Comment by tobyplowy** [comment]
@Ferk Look at [this picture](http://m.imgur.com/xDbpV6P?r) for reference It's a Minecraft skin there are textures for the left leg and the right leg and they are basically the same texture but they both have their own napping so if you want something different left side you can (Every bodypart needs Its Own Texture on the mapping no faces using the same textures on the texture map every face needs its own texture on the texture map) even if two sides use the same texture it would just be a copy paste texture but on a different face (all 6 faces Need a Unique Texture on the Texture map and all body parts are on One image map) I hope this makes Sense Now I know it a lot Of Stuff

**Comment by Ferk** [comment]
@tobyplowy I didn't talk about different files. When I said "one texture for each of the 8 legs" I didn't mean one "texture file" but more like one specific texture strip (inside the same file), vs using the same texture data for all the legs.

**Comment by tobyplowy** [comment]
@Ferk sorry for the misunderstanding 

I'm talking about Every Bodypart on one image no Different images in one image file just one simple image
Also Can the model be organised Like Minecraft textures? Where is does the six faces for every bodypart
Like this http://m.imgur.com/xDbpV6P?r

**Comment by Ferk** [comment]
@tobyplowy Yes, that's the same I was talking about, I know you didn't mean different different texture files, you meant different textures in the same image.
And actually Minecraft does not have all the 8 legs of its spiders in the same texture file. There's only textures for 1 leg, and it just gets reused.

But anyway, I think this is off-topic. The way the model is UV mapped is the least of the problems regarding "Mob's animals/monsters", imho.

**Comment by tobyplowy** [comment]
@Ferk the texture layout is really important but enough about that, if we add mobs we better do it good on the first try, I think we should begin thinking about what monsters and animals we want (just being Curious here What would you like to see ferk?

**Comment by paramat** [comment]
Personally i would like to see re-addition of classic MT mobs, certainly the Oerkki, perhaps with a new simpler model, although the SImple Mobs version has a good suitable model design. Maybe Dungeon Master too.

**Comment by tobyplowy** [comment]
@paramat I found this topic that May be interesting https://forum.minetest.net/viewtopic.php?id=1594 It has some interesting models (Also The Dungeon master should look like this http://rs288.pbsrc.com/albums/ll190/melkor_666/DM.jpg?w=480&h=480&fit=clip

**Comment by tobyplowy** [comment]
@paramat I think zombies and mummys shuld get added also some marine animals like shark dolphin and normal fish, maybe also some farm animals like chicken sheep cow horse and donkey (and I really want lama's)

**Comment by tobyplowy** [comment]
@paramat I think zombies and mummys shuld get added also some marine animals like shark dolphin and normal fish, maybe also some farm animals like chicken sheep cow horse and donkey (and I really want lama's) Bdw I really Wonder what @C1ffisme @rubenwardy @DonBatman @tenplus1 @sofar would like to see

**Comment by NathanSalapat** [comment]
@tobyplowy More gets done when you create models with the UV layouts you'd like, than when you just talk about them. ;)

**Comment by tobyplowy** [comment]
@NathanSalapat I Don't really get that, Could you Explain It a bit Better Sorry :/

**Comment by tenplus1** [comment]
@tobyplowy - I think what Nathan means is that it's all good and well wanting something and writing it down, but it would be even better if you could source actual models, textures and the like to make it possible.

**Comment by sofar** [comment]
OT: (this thread is wildly out of bounds already)

I've stated my opinions in several threads on the forums, but it can be summarized as:

For MTG, only 4-5 monsters total - 2 farm animals, 1 aggressive surface monster, 2 cave dwellers. Extremely well modeled and textured, proper animations. Pluggable AI and state transitions. Mobs are permanent, spawned by perlin noise, and can handle /clearobjects (a block reload should respawn them).

I think almost all of the MT mobs I've seen look like crap, and the code that drives them is unmaintainable. I don't ever want that to go into mtg.

Perhaps my standards are unreasonably high.

**Comment by tobyplowy** [comment]
@sofar I Wouldn't say They are crap (I Would call them not good it sounds nicer but it's the same thing) You got to give Them some credit after all do it for free ( your standards are not that High This is what good Quality mobs are) the only thing I Want Is that the Texture is easy to edit and dat You Don't Need A model to know where the textures are located (and Every mob should have a organized system for the textures)

**Comment by paramat** [comment]
> I think almost all of the MT mobs I've seen look like crap

I agree, and you have a good approach to modelling, your sheep are suitably designed.

**Comment by NathanSalapat** [comment]
Agree with @tenplus1 and @sofar.
If you have a design in mind make it, and this should be closed, forums are the place for discussion.

**Comment by tobyplowy** [comment]
@NathanSalapat jeh but if we do that It's Never Gonna happen

**Comment by C1ffisme** [comment]
Do we want to concentrate our efforts on making lua mobs in MTG and work to make the engine support mobs made in that way, or making a Mob API in the engine for (potentially) faster mobs? I'm fine either way, but I feel like making this decision would be good before starting the process of designing mobs for Minetest.

We can idealize all kinds of mobs, or wish that the old ones would return, but I think that making a good template or API for developers and modders alike would be a good first step.

**Comment by sofar** [comment]
https://forum.minetest.net/viewtopic.php?f=9&t=15572

**Comment by rubenwardy** [comment]
I would personally like lots of realistic animals and a few monsters. Mobs make an environment feel alive.

Wolves, bears, rats, sheep, chickens, cows, various birds and fishes
Along with classic Mt 0.3 monsters

Really mobs should be engine supported, so the technical discussion should be there in the issue there. I think it's okay to discuss game design here for mtg, but talking about skins is just an irrelevant detail

**Comment by tobyplowy** [comment]
@C1ffisme I think development should go to performance because that's one of the big problems with mob mods out there (they take all the performance)

**Comment by tobyplowy** [comment]
Sorry about that

**Comment by tobyplowy** [comment]
@rubenwardy I agree with everything you just Said :)

**Comment by paramat** [comment]
> jeh but if we do that It's Never Gonna happen

Whether mobs get added does not depend on the presence of an issue, this is always in the back of devs minds. And the intention is already in #515 
This thread now has some decent discussion in it so can be kept open i think, 515 is more of an overview and isn't very focussed. Similarly for example i think we will need a new issue focussing in detail on the addition of carts.

I agree with sofar that for MTGame mobs should be kept simple and few in number. The aggressive surface monster could be the Oerkki. tobyplowy and rubenwardy's long lists of mobs is too much for the basic game, keep that for a 2nd subgame.

**Comment by tobyplowy** [comment]
@paramat I know this is off topic but I have to mention it 

> keep that for a 2nd subgame.

Also you are right we shouldn't over-complicate the best Minecraft game
But I want a official more Advanced game

**Comment by tobyplowy** [comment]
@paramat I Just Found this, i really like The look Maybe They Could Be used as Inspiration http://techne.zeux.me/profile/Deadboy/models they look very realistic but Still blocky

**Comment by C1ffisme** [comment]
I decided for the sake of it, I'd clean up the grammar and spelling in the original post for better clarity and reasoning. @tobyplowy If you feel like copying and pasting, that's okay with me. (Although, don't copy that note. That might be a little confusing.)

```
I think Minetest Game should have mobs or animals (because playing survival without mobs is kind of boring). There is also no sense of danger or fear without mobs, the only way you can die is fall damage, cactus or lava. Also, I know there are mods for this, I'm just saying because I know the first few comments will point that out. (**Note to @tobyplowy: That last sentence might not be needed anymore, since the "first few comments" have already been posted. ;)** ) XD (This could be really good for a proper mob API because the ones mods use are laggy, or are filled with unnecessary features.)

Also I think there should be a setting for mobs if people don't want them or if their systems cannot handle them.
```

**Comment by tobyplowy** [comment]
@C1ffisme Sorry I'm Really Trying to do my best I'm not a native English speaking person :/

**Comment by C1ffisme** [comment]
@tobyplowy I know, that's why I decided to give you some help. :wink:

**Comment by paramat** [comment]
Those models are far too complex.

**Comment by tobyplowy** [comment]
@paramat @C1ffisme @NathanSalapat @tenplus1 @sofar I Made this Model for a Pig do you like it? (if you like it I could Give You A download and improve The Texture)
https://sketchfab.com/models/bb78d795f47d4165b52494e7ce1a9d59

**Comment by rubenwardy** [comment]
Models are not the problem at the moment. The problem is how this feature will be implemented, not what specific animals will be present or what they will look like.

**Comment by tobyplowy** [comment]
@rubenwardy oh ok sorry :/

**Comment by tobyplowy** [comment]
You know what will be useful if Minetest wuld have a engine with is like Minecraft
In with mobs are made Out Of code and not models

**Comment by C1ffisme** [comment]
@tobyplowy Cool model, though. I like it.

**Comment by tobyplowy** [comment]
@C1ffisme Thanks

**Comment by tobyplowy** [comment]
@C1ffisme although It Looks alot Like The Pig from Minecraft that was not intentional btw :/

**Comment by godisgood4** [comment]
I would absolutely love the addition of mobs into Minetest Game, but only with the addition of a difficulty slider in the pause menu (not in a configuration file).

**Comment by octacian** [comment]
I would like to see _realistic_ mobs in MT. Not realistic as in shape, but as in what is added. Stuff like a lion, fisher, or cheetah. They could be hostile too. Peaceful mobs like horses and dogs could also be added. Monsters on the other hand, no no. That's why I do not play MC, I do not like the mythical monsters and witchcraft aspects of it.

**Comment by tobyplowy** [comment]
@octacian Thet ruins The fun ;)

**Comment by octacian** [comment]
No, @tobyplowy, it does not ruin the fun. Animals (note animals, not monsters) like bears, worlverines, and lions could attack players. Wolfs too, and coyotes when angered. Without monsters, there could/would be just as much "fun" even if all the creatures were literal animals.

Really though, if monsters get added, I'm outta here. I'll finally start working on my own game then :)

**Comment by tobyplowy** [comment]
@octacian he is not smiling thats the shadow of his Nose XD 
Also it's really hard to make it better because with the cubes you can't really make it over detaild without having a ton of cubes and that wuld overdue the simplistic Style

Also a little reminder I Made This model with Google SketchUp :/
https://sketchfab.com/models/bb78d795f47d4165b52494e7ce1a9d59

**Comment by rubenwardy** [comment]
As I've said, I'm in support of realistic animals. Although I'd also like to see Dms and classic 0.3 mobs for nostalgia reasons

**Comment by tobyplowy** [comment]
@octacian @rubenwardy I'm good with the fact that there is no monster mobs I Just Want One Monster And that is The Dungeon after all

**Comment by Ferk** [comment]
Probably the difficulty setting should rather be a value to control the chance for the monsters to spawn. So they can be still enabled by default but be very rare, while allowing users to make them absurdly common, or turn them off completely if they really want to.

Maybe there could be a separate setting for animals to control their population too. But since animals might need to be more persistent (for them to be farmable), that one might have to be more of a mapgen setting than something that can be changed on the fly.

**Comment by sofar** [comment]
This discussion is getting terribly out of hand.

I don't think this feature request itself is ever going to get rejected. What *is* going to get rejected is an implementation (code) that does the wrong thing.

But, until an actual code submission is done, putting additional requests or requirements in this feature request is only going to dilute it.

This "feature request" should have had 2 comments:
1) The original top OP comment stating "Minetest needs mobs"
2) The comment where the feature request is "approved". (Alas, we don't have such a thing for requests).

All the other comments are entirely moot and we'll have the same drawn out discussion when someone actually submits some code.

so, I suggest that everyone stops feeding the suggestion box here and either *code* it or go play minetest.

I suggest we lock this topic. We're all in agreement that minetest needs mobs, I don't need notifications about that anymore.

**Comment by octacian** [comment]
I agree with @sofar. I will begin to write a new mod called "mobs," which I will develop with the idea of merging it with MTG. When I get it going, I will post a link. Unless others have better ideas, this is what I've got. I don't know how it's going to turn out, but one thing IK for sure, is that I'm def gonna need lots of input from others :D

**Comment by tobyplowy** [comment]
@octacian I'm sensing sarcasm here but I'm not sure ( are you for real or is it a joke?)

**Comment by octacian** [comment]
@tobyplowy for real... How long did it take me to make those 3D torches?

On the other hand though, there is not a chance I'm gonna do it or be able to without help. I'm more than willing to put in the effort and make this happen, but I don't have much time for MT ATM. If somebody gets to it before me, I'll honestly be glad. I'll still contribute whenever possible, but, you get it.

**Comment by paramat** [comment]
I agree with sofar, this sort of vague discussion belongs in the forum, we already have an issue for this #515 and it's not going to be forgotten.

**Comment by tobyplowy** [comment]
@octacian please do that (also It Would Be Better if its programd in the engine And Not as a mod (for performance and possibility reasons) also it would be greatly appreciated if you could do this I'm sure you will get a lot of help :)

**Comment by tobyplowy** [comment]
I guess a good mob ai will never get added :/

**Comment by sofar** [comment]
Just come help me with entity_ai and see if we can make a good API  and mob code design. The skeleton is already usable.

**Comment by tobyplowy** [comment]
@sofar If Only I could Help But i cannot program sadly :/

**Comment by tobyplowy** [comment]
@sofar someone just has to break the ice once that happened development on the mob ai will skyrocket

**Comment by paramat** [comment]
There's a very high chance optional mobs will be added, so don't worry. Now please discuss elsewhere.
